### PR TITLE
Add functions to `list`

### PR DESCRIPTION
**Pull request by samuelchassot**

### CODE REVIEW

File: pyfuncol/list.py [hunk 1]
```diff
@@ -158,6 +158,107 @@ def index_of(self: List[A], elem: A) -> int:
             return i
     return -1
 
+def fold_left(self: List[A], z: B, op: Callable[[B, A], B]) -> B:
+    """
+    Applies a binary operator to a start value and all elements of this sequence, going left to right.
+
+    Args:
+        z: start value
```
Comment (Gondolav) [inline pyfuncol/list.py:166]: Please capitalise each sentence and put a period at the end ;)

File: pyfuncol/list.py [hunk 2]
```diff
@@ -158,6 +158,107 @@ def index_of(self: List[A], elem: A) -> int:
...
+        acc = op(x, acc)
+    return acc
+
+def forall(self: List[A], p: Callable[[A], bool]) -> bool:
+    """
+    Tests whether a predicate holds for all elements of this list.
+
+    Args:
+        p: the predicate used to test elements.
+        returns: true if this list is empty or the given predicate p holds for all elements of this list, otherwise false.
```
Comment (Gondolav) [inline pyfuncol/list.py:204]: You should write Returns: like before

File: pyfuncol/list.py [hunk 3]
```diff
@@ -158,6 +158,107 @@ def index_of(self: List[A], elem: A) -> int:
...
+    
+    return True
+
+def head(self: List[A]) -> A:
+    """
+    Selects the first element of this iterable collection.
+
+    Note: might return different results for different runs, unless the underlying collection type is ordered.
+
+    Exceptions thrown:
```
Comment (Gondolav) [inline pyfuncol/list.py:218]: You should use instead `Raises:` as it is understood by the docs when they autogenerate the API reference

File: pyfuncol/list.py [hunk 4]
```diff
@@ -158,6 +158,107 @@ def index_of(self: List[A], elem: A) -> int:
...
+    if n < 0:
+        return []
+    if len(self) <= n:
+        return self
+    
+    return self[0:n]
+
+def length(self: List[A]) -> int:
+    """
+    The length (number of elements) of the list. size is an alias for length in Seq collections.
```
Comment (Gondolav) [inline pyfuncol/list.py:256]:
```suggestion
    Returns the length (number of elements) of the list. `size` is an alias for length.
```

File: pyfuncol/tests/test_list.py [hunk 5]
```diff
@@ -57,3 +57,77 @@ def test_foreach():
...
+    a = l.fold_right(0, lambda n, acc: acc + n)
+    assert a == 6
+
+def test_fold_right_concat():
+    a = l.fold_right("", lambda n, acc: acc + str(n))
+    assert a == "321"
+
+def test_forall_gt_zero():
+    a = l.forall(lambda n: n > 0)
+    assert a
```
Comment (Gondolav) [inline pyfuncol/tests/test_list.py:80]: Specify explicitly == True here ;)

File: pyfuncol/tests/test_list.py [hunk 6]
```diff
@@ -57,3 +57,77 @@ def test_foreach():
...
+def test_head():
+    h = l.head()
+    assert h == 1
+
+def test_head_empty():
+    l = []
+    flag = False
+    try:
+        l.head()
+    except IndexError:
```
Comment (Gondolav) [inline pyfuncol/tests/test_list.py:95]: You can use:

```python
with pytest.raises(IndexError):
        l.head()
```

File: pyfuncol/tests/test_list.py [hunk 7]
```diff
@@ -57,3 +57,77 @@ def test_foreach():
...
+    a = l.take(4)
+    assert a == l
+
+def test_take_smaller_len():
+    a = l.take(2)
+    assert a == [1,2]
+
+def test_length():
+    a = l.length()
+    assert a == 3
```
Comment (Gondolav) [inline pyfuncol/tests/test_list.py:133]: Maybe you can also add an assert that `l.length() == l.size()`

**Review by Gondolav — CHANGES_REQUESTED** [review]
Great work, just a few things to adjust, thanks!

File: pyfuncol/list.py (hunk 2 above)
Comment (samuelchassot) [reply pyfuncol/list.py:204]: 👍

File: pyfuncol/list.py (hunk 3 above)
Comment (samuelchassot) [reply pyfuncol/list.py:218]: 👍

File: pyfuncol/tests/test_list.py (hunk 6 above)
Comment (samuelchassot) [reply pyfuncol/tests/test_list.py:95]: thanks :)

File: pyfuncol/tests/test_list.py (hunk 7 above)
Comment (samuelchassot) [reply pyfuncol/tests/test_list.py:133]: good idea :)

**Comment by Gondolav** [comment]
That's great! Just run `black` to reformat everything and we can merge!

**Comment by Gondolav** [comment]
I forgot, we should either update the API section of the README or simply remove the listings and just put the link to the API reference. WDYT?

**Comment by samuelchassot** [comment]
IMHO it is better to avoid having duplicated information to update, it is too much prone to error. I would say that we either only link to the API (we can even have one link for each class like `list`, `dict`, ... to be even quicker to the information) or automating the process of updating the README. The former is probably the sensible option ;)

**Comment by Gondolav** [comment]
Yes I agree, then let's remove them and link directly the API doc for `list` and `dict`.

**Comment by samuelchassot** [comment]
I merged main and update the README

File: README.md [hunk 8]
```diff
@@ -39,11 +39,11 @@ import pyfuncol
 
 ### API
 
-For lists: `map`, `filter`, `flat_map`, `flatten`, `contains`, `distinct`, `foreach`, `group_by`, `is_empty`, `size`, `find`, `index_of`.
+For lists, please refer to this [API reference](https://pyfuncol.readthedocs.io/en/latest/pyfuncol.html#module-pyfuncol.list)
```
Comment (Gondolav) [inline README.md:42]:
```suggestion
For lists, please refer to the [docs](https://pyfuncol.readthedocs.io/en/latest/pyfuncol.html#module-pyfuncol.list)
```

File: README.md [hunk 9]
```diff
@@ -39,11 +39,11 @@ import pyfuncol
...
 
-For dictionaries: `map`, `filter`, `flat_map`, `contains`, `foreach`, `is_empty`, `size`, `to_list`.
+For dictionaries, please refer to this [API reference](https://pyfuncol.readthedocs.io/en/latest/pyfuncol.html#module-pyfuncol.dict)
```
Comment (Gondolav) [inline README.md:44]:
```suggestion
For dictionaries, please refer to the [docs](https://pyfuncol.readthedocs.io/en/latest/pyfuncol.html#module-pyfuncol.dict)
```

File: README.md [hunk 10]
```diff
@@ -39,11 +39,11 @@ import pyfuncol
...
 
-For more details, look at the [API reference](https://pyfuncol.readthedocs.io/en/latest/modules.html).
+For more details, please have a look to [API reference](https://pyfuncol.readthedocs.io/en/latest/modules.html).
```
Comment (Gondolav) [inline README.md:46]:
```suggestion
For more details, please have a look at the [API reference](https://pyfuncol.readthedocs.io/en/latest/modules.html).
```

**Comment by samuelchassot** [comment]
👍

**Comment by Gondolav** [comment]
Great, let's merge once the checks have passed!

**Review by Gondolav — APPROVED** [review]

**Pull request merged**